libs/orderbook/src/lib/orderbook-data.ts

```diff
@@ -0,0 +1,310 @@
+import produce from 'immer';
+import groupBy from 'lodash/groupBy';
+import type {
+  MarketDepth_market_depth_sell,
+  MarketDepth_market_depth_buy,
+} from './__generated__/MarketDepth';
+import type {
+  MarketDepthSubscription_marketDepthUpdate_sell,
+  MarketDepthSubscription_marketDepthUpdate_buy,
+} from './__generated__/MarketDepthSubscription';
+
+export interface CumulativeVol {
+  bid: number;
+  relativeBid?: number;
+  ask: number;
+  relativeAsk?: number;
+}
+
+export interface OrderbookData {
+  price: number;
+  bidVol: number;
+  bidVolByLevel?: Record<number, number>;
+  relativeBidVol?: number;
+  askVol: number;
+  askVolByLevel?: Record<number, number>;
+  relativeAskVol?: number;
+  cumulativeVol: CumulativeVol;
+}
+
+const getGroupPrice = (price: number, resolution: number) =>
+  Math.round(price / resolution) * resolution;
+
+const maxVolumes = (orderbookData: OrderbookData[]) => {
+  let bidVol = 0;
+  let askVol = 0;
+  let cumulativeVol = 0;
+  orderbookData.forEach((data) => {
+    bidVol = Math.max(bidVol, data.bidVol);
+    askVol = Math.max(askVol, data.askVol);
+  });
+  cumulativeVol = Math.max(
+    orderbookData[0]?.cumulativeVol.ask,
+    orderbookData[orderbookData.length - 1]?.cumulativeVol.bid
+  );
+  return {
+    bidVol,
+    askVol,
+    cumulativeVol,
+  };
+};
+
+const toPercentValue = (value?: number) => Math.ceil((value ?? 0) * 100);
+
+const updateRelativeData = (data: OrderbookData[]) => {
+  const { bidVol, askVol, cumulativeVol } = maxVolumes(data);
+  data.forEach((data) => {
+    data.relativeAskVol = toPercentValue(data.askVol / askVol);
+    data.relativeBidVol = toPercentValue(data.bidVol / bidVol);
+    data.cumulativeVol.relativeAsk = toPercentValue(
+      data.cumulativeVol.ask / cumulativeVol
+    );
+    data.cumulativeVol.relativeBid = toPercentValue(
+      data.cumulativeVol.bid / cumulativeVol
+    );
+  });
+};
+
+export const compact = (
```

Comment:
chore (blocking): Could really use some comments at the top of these functions outlining whats happening